File: leet_DP_0377_combination_sum_IV.java
class Solution {
    public int combinationSum4(int[] nums, int target)
    {
        int[]   dp = new int[target + 1];
        int     i, j;

        Arrays.fill(dp, 0);
        dp[0] = 1;
        i = 0;
        while (++i < target + 1)
        {
            j = -1;
            while (++j < nums.length)
            {
                if (i - nums[j] > -1)
                    dp[i] += dp[i - nums[j]];
            }
        }
        return dp[target];
    }
}

File: leet_DP_0377_combination_sum_IV.scala
object Solution {
    def combinationSum4(nums: Array[Int], target: Int): Int = {
        // var dp = new Array.fill[Int](target + 1)(0);
        var dp = Array.fill[Int](target + 1)(0);
        dp(0) = 1;
        for (i <- 1 to target) // for loop in scala: upper limit is included
        {
            for (n <- nums)
            {
                if (i - n > -1)
                {
                    dp(i) += dp(i - n);
                }
            }
        }
        dp(target)
    }
}
